src/main/java/net/neoforged/neoforge/registries/DeferredRegister.java

```diff
@@ -5,319 +5,267 @@
-    public <I extends T> RegistryObject<I> register(final String name, final Supplier<? extends I> sup) {
+    public <I extends T> DeferredHolder<T, I> register(final String name, final Function<ResourceLocation, ? extends I> func) {
         if (seenRegisterEvent)
             throw new IllegalStateException("Cannot register new entries to DeferredRegister after RegisterEvent has been fired.");
         Objects.requireNonNull(name);
-        Objects.requireNonNull(sup);
-        final ResourceLocation key = new ResourceLocation(modid, name);
-
-        RegistryObject<I> ret;
-        if (this.registryKey != null)
-            ret = this.optionalRegistry
-                    ? RegistryObject.createOptional(key, this.registryKey, this.modid)
-                    : RegistryObject.create(key, this.registryKey, this.modid);
-        else
-            throw new IllegalStateException("Could not create RegistryObject in DeferredRegister");
-
-        if (entries.putIfAbsent((RegistryObject<T>) ret, sup) != null) {
+        Objects.requireNonNull(func);
+        final ResourceLocation key = new ResourceLocation(namespace, name);
+
+        DeferredHolder<T, I> ret = createHolder(this.registryKey, key);
+
+        if (entries.putIfAbsent(ret, () -> func.apply(key)) != null) {
             throw new IllegalArgumentException("Duplicate registration " + name);
         }
 
         return ret;
     }
 
     /**
-     * Only used for custom registries to fill the forge registry held in this DeferredRegister.
-     *
-     * Calls {@link RegistryBuilder#setName} automatically.
-     *
-     * @param sup Supplier of a RegistryBuilder that initializes a {@link IForgeRegistry} during the {@link NewRegistryEvent} event
-     * @return A supplier of the {@link IForgeRegistry} created by the builder.
-     *         Will always return null until after the {@link NewRegistryEvent} event fires.
+     * Create a {@link DeferredHolder} or an inheriting type to be stored.
+     * 
+     * @param registryKey The key of the registry.
+     * @param key         The resource location of the entry.
+     * @return The new instance of {@link DeferredHolder} or an inheriting type.
+     * @param <I> The specific type of the entry.
      */
-    public Supplier<IForgeRegistry<T>> makeRegistry(final Supplier<RegistryBuilder<T>> sup) {
-        return makeRegistry(this.registryKey.location(), sup);
+    protected <I extends T> DeferredHolder<T, I> createHolder(ResourceKey<? extends Registry<T>> registryKey, ResourceLocation key) {
+        return DeferredHolder.create(registryKey, key);
     }
 
     /**
-     * Creates a tag key based on the current modid and provided path as the location and the registry name linked to this DeferredRegister.
-     * To control the namespace, use {@link #createTagKey(ResourceLocation)}.
+     * This method is used to configure a custom modded registry. It can only be invoked by a single DeferredRegister instance for a given registry key.
```

Comment:
Strengthen the requirement described in that second sentence.
```suggestion
     * This method is used to configure a custom modded registry. It must only be invoked by a single DeferredRegister instance for a given registry key.
```

Reply:
Technically modders are not forced to use DR to create a registry, hence the suggestive wording.

Reply:
Hmm. I prefer using "must" (since as I understand it, using "must" doesn't require the modder to call it), but I see how it can be interpreted as that.